cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,461 @@
+			slog.Error("processLiveCaptionsForTrack: we have zero segments in the window. Should not be possible.",
+				slog.String("trackID", trackID))
+			break
+		}
+
+		var oldestSegment segmentSamples
+		oldestSegment, segments = segments[0], segments[1:]
+		var cutUpTo int
+		if len(segments) == 0 {
+			// We don't have a complete next segment yet: cut to end of oldest segment.
+			cutUpTo = oldestSegment.End
+		} else {
+			// Cut up to start of segment we're keeping.
+			cutUpTo = segments[0].Start
+		}
+		if cutUpTo > len(window) {
+			// Don't panic, defensive, shouldn't happen.
+			cutUpTo = len(window)
+		}
+		window = window[cutUpTo:]
+
+		// Adjust our marker for where we've transcribed.
+		// e.g., prevTranscribedPos was 10, we've cut 6, new pos is 10 - 6 = 4.
+		prevTranscribedPos -= cutUpTo
+	}
+
+	return window, prevTranscribedPos
+}
+
+func (t *Transcriber) startTranscriberPool() {
+	for i := 0; i < t.cfg.LiveCaptionsNumTranscribers; i++ {
+		t.transcriberWg.Add(1)
+		go t.handleTranscriptionRequests(i)
+	}
+}
+
+func (t *Transcriber) handleTranscriptionRequests(num int) {
+	slog.Debug(fmt.Sprintf("live captions, handleTranscriptionRequests: starting transcriber #%d", num))
+
+	transcriber, err := t.newLiveCaptionsTranscriber()
+	if err != nil {
+		slog.Error("live captions, handleTranscriptionRequests: failed to create transcriber",
+			slog.String("err", err.Error()))
+		return
+	}
+	defer func() {
+		err := transcriber.Destroy()
+		if err != nil {
+			slog.Error("live captions, handleTranscriptionRequests: failed to destroy transcriber",
+				slog.String("err", err.Error()))
+		}
+		t.transcriberWg.Done()
+	}()
+
+	for {
+		select {
+		case <-t.transcriberDoneCh:
+			slog.Debug(fmt.Sprintf("live captions, handleTranscriptionRequests: closing transcriber #%d", num))
+			return
+		case packet := <-t.transcriberQueueCh:
+			transcribed, _, err := transcriber.Transcribe(packet.pcm)
+			if err != nil {
+				slog.Error("live captions, handleTranscriptionRequests: failed to transcribe audio samples",
+					slog.String("err", err.Error()))
+				packet.retCh <- ""
+				return
+			}
+
+			if len(transcribed) == 0 {
+				packet.retCh <- ""
+			} else {
+				packet.retCh <- transcribed[0].Text
+			}
+		}
+	}
+}
+
+func (t *Transcriber) newLiveCaptionsTranscriber() (transcribe.Transcriber, error) {
+	switch t.cfg.TranscribeAPI {
+	case config.TranscribeAPIWhisperCPP:
+		return whisper.NewContext(whisper.Config{
+			ModelFile:     filepath.Join(getModelsDir(), fmt.Sprintf("ggml-%s.bin", string(t.cfg.LiveCaptionsModelSize))),
+			NumThreads:    t.cfg.LiveCaptionsNumThreadsPerTranscriber,
+			NoContext:     true, // do not use previous translations as context for next translation: https://github.com/ggerganov/whisper.cpp/pull/141#issuecomment-1321225563
+			AudioContext:  512,  // a bit more than 10seconds: https://github.com/ggerganov/whisper.cpp/pull/141#issuecomment-1321230379
+			PrintProgress: false,
+			Language:      "en",
```

Comment:
How hard would it be to come up with something dynamic? Like autodetecting the first few seconds and then use the resulting lang code throughout, regardless of which track it's coming from. I think it's fair to assume same call == same language.

Otherwise I feel we may need to export it so that it can at least be controlled by admins.

Reply:
I'm hesitant to try to detect it, considering the trouble we've had with that in the past, and we'd need rules (what confidence, where in the meeting do we detect, for how long, etc.), and then there's the issue of meetings that start with one language and go to the main language later. If we're going to do a single set language, probably we should make it per-call configurable, or system-level (perhaps in the system-console).
Needs UX, so made a ticket: https://mattermost.atlassian.net/browse/MM-57408

Reply:
Yes, I was assuming system console (global setting) which wouldn't need much UX work, either a dropdown or a simple text field expecting the country code. I'd like to avoid the hard-coded `en` since we already support autodetection for post call transcripts.

Reply:
Cool, added the config setting

Reply:
Done in https://github.com/mattermost/mattermost-plugin-calls/pull/633/commits/80602b99a142f3b886d9b929db09f025d2d8800f